### ISSUE TITLE
How to use ImGui rounding style for ImGui::MenuItems?

### ISSUE THREAD

**Issue by SrinSS01** [ocornut/imgui#7589]
### Version/Branch of Dear ImGui:

Version 1.90.6, Branch: master

### Back-ends:

imgui_impl_XXX.cpp + imgui_impl_XXX.cpp

### Compiler, OS:

Windows 11 + emscripten GCC

### Full config/build information:

![image](https://github.com/ocornut/imgui/assets/60167747/4e51ad3a-8cf3-41c5-a3fd-3e8a3a4a5db1)
![image](https://github.com/ocornut/imgui/assets/60167747/a09f3461-170c-44a6-9ebe-771945d2a2cc)

### Details:

**How to use ImGui rounding style for ImGui::MenuItems?**

I am using ImGui in web browser through emscripten.
while searching through the rounding style vars, the only ones available are
- ImGuiStyleVar_WindowRounding
- ImGuiStyleVar_ChildRounding
- ImGuiStyleVar_PopupRounding
- ImGuiStyleVar_FrameRounding
- ImGuiStyleVar_ScrollbarRounding
- ImGuiStyleVar_GrabRounding
- ImGuiStyleVar_TabRounding

I was rendering an ImGui Menu item and was wondering how to enable rounding effect for each Items
The popup window is having rounded corners but the items inside it are still sharp edges which makes it look kinda odd, you can refer to the attached screenshot.

I tried all the style vars but none are showing any effects.

Any idea how to make rounded Menu items?

### Screenshots/Video:

![image](https://github.com/ocornut/imgui/assets/60167747/06b00d72-2405-4741-8374-c3402e90ca35)

### Minimal, Complete and Verifiable Example code:

```cpp
// Here's some code anyone can copy and paste to reproduce your issue
ImGuiStyle& style = ImGui::GetStyle();
        
  style.PopupRounding = 9;
  if (ImGui::BeginMainMenuBar()) {
      if (ImGui::BeginMenu("File")) {
          ImGui::MenuItem("Test", "Ctrl+T");
          ImGui::MenuItem("Test", "Ctrl+T");
          ImGui::MenuItem("Test", "Ctrl+T");
          ImGui::MenuItem("Test", "Ctrl+T");
          ImGui::EndMenu();
      }
      ImGui::EndMainMenuBar();
  }
```

**Comment by ocornut** [comment]
We don't support rounding Menu Items or Selectables.
One reason is Selectables are meant to fill space without gaps. Another reason is that I have an idea in mind for a future more general-purpose Selectable API which may make it more difficult to implement if rounded.

I suppose menu items, even though they are selectable, may not need to cater for those restrictions and we could lift them for Menu Items. However if your aim is to make everything round you are not going to be able to do it with selectables.

> Full config/build information:

(FYI, there is a "Copy to clipboard" button in that About window.)

**Comment by SrinSS01** [comment]
Thanks for the update. I don't require round effect to selectables in general, I would appreciate if only Menu items can have that feature. 

> (FYI, there is a "Copy to clipboard" button in that About window.)

In ImGui glfw opengl3 implementation with emscripten, `copy to clipboard` doesn't seem to be working from the browser for me.

**Comment by LxeC** [comment]
I also just came across this limitation and need it for my project. Menu items should be able to be rounded (could also be done individually from other selectables). The rounding adds so much more professional look in such a setup. 

I'd also like to ask if it'd be possible to have the option to have rounded look for menu bar items as well when they are highlighted. This is trivial to do in CSS but I want to use imgui.

**Comment by janitor-raus** [comment]
I found myself noticing this situation as well, where the Selectables aren't roundable, because they're explicitly set to 0.0f rounding, not tied to any rounding variables.
I ended up making an addition to my cmake pipeline to allow automatic application of patch diffs to the fetchcontent, so now at least I can both apply customizations retroactively and distribute them with my project without making forks or awkward replacement files.

<img width="1221" height="85" alt="Image" src="https://github.com/user-attachments/assets/94ce130b-ceec-4cfc-acfb-c0e061289165" />

**Comment by ocornut** [comment]
I explicitly answered above that I have a generalization of Selectable planned that I don’t know how to make work with rounding.

**Comment by ocornut** [comment]
It also doesn’t make much visual sense with multi-selection IMHO. But probably MenuItem() should allow it.

**Comment by janitor-raus** [comment]
> It also doesn’t make much visual sense with multi-selection IMHO. But probably MenuItem() should allow it.

Yeah in some cases it indeed will look a bit weird. Particularly for the 2D layouts. I tested a bit in the demo window:

This one looks fine, though the grid would look better if it were outlined or forced square:
<img width="350" alt="Image" src="https://github.com/user-attachments/assets/9ae196a1-62f0-4897-834f-c3915bd89ef3" />

This one's admittedly better forced to be explicit squares, rather than have rounding:
<img width="400" alt="Image" src="https://github.com/user-attachments/assets/0f6bba9b-7dae-4009-9fcc-97d2291e47a4" />

Multi-select in trees seems to be a unique case, as my edit does not affect it:
<img width="300" alt="Image" src="https://github.com/user-attachments/assets/e0a8b937-8272-471f-970a-49a5b5c46896" />

These also look fine to me rounded, though the primary selection outline isn't rounding. I might also patch that on my side for consistency:
<img width="300" alt="Image" src="https://github.com/user-attachments/assets/090d21b6-d56a-4fec-8806-5c28b4a3922e" />

List boxes are also alright:
<img width="450" alt="Image" src="https://github.com/user-attachments/assets/321439e9-32c0-4d11-9d71-8f8d26e18fb6" />

From a quick look, I'd say it's not all _that_ bad. I can definitely live with it, and if there's some selectable section that shouldn't have rounding, I could be explicit about it and push a 0 to the style. I personally feel it's acceptable as a default case, but I've also no problem applying little patches to my project for preference :)

**Comment by ChiefyChief23** [comment]
Hello all, long time listener first time caller

So as @ocornut kindly mentioned on this thread I made a PR to get rounding on selectables. It is a very simple change as you can see from the PR. The implementation adds an additional element to the style similar to `FrameRounding`/`WindowRounding` so worst case scenario now could be, follow a Push/Pop style var so you could have rounding for menu items and then Push/Pop filled for multi-select as mentioned above. Completely up the developer.

@janitor-raus The reason tree nodes are not affected by the rounding is because they are rendering from a different function than the selectable. You can find it within `bool ImGui::TreeNodeBehavior(ImGuiID id, ImGuiTreeNodeFlags flags, const char* label, const char* label_end)` towards the end with `RenderFrame` like in the selectables.

```cpp
            // Unframed typed for tree nodes
            if (hovered || selected)
            {
                const ImU32 bg_col = GetColorU32((held && hovered) ? ImGuiCol_HeaderActive : hovered ? ImGuiCol_HeaderHovered : ImGuiCol_Header);
                RenderFrame(frame_bb.Min, frame_bb.Max, bg_col, false);
            }
```

Similar approach could be made with the tree nodes for rounding by adding an additional element for it but that's up for discussion as I don't have the big picture view as some others have.

**Comment by janitor-raus** [comment]
@ChiefyChief23 I am aware. I dug into the code some more post-comment with the pictures, and as far as I can tell, I've managed to patch all the important parts? Tree-specific RenderFrame() call aside, it turned out I had to remove the NoRounding flag for the Nav cursor render for consistency of selection outlines (square vs the rounded frame). Actual patch for reference here: https://github.com/janitor-raus/CubeChip/blob/master/patches/imgui/01_selectable_rounding.patch

As for downsides.. I have yet to find anything misbehaving **yet**. Not guaranteeing this won't introduce weirdness somewhere or that I didn't miss something else would also need the rounding touch-up, but from a quick look-thru the demo window and a bunch of navigating with the keyboard, it seems fine.

In any case, that pull request of yours.. could take my patch one step further. I relied on FrameRounding as you can see in the patch. Felt that I didn't have to do add a new *Rounding entry, as typically anything that has a frame is already using FrameRounding, thus I didn't see a reason to introduce a layer of mismatch between them.

Example pics:

Nav cursor over a typical field:
<img alt="Image" src="https://github.com/user-attachments/assets/fad2f730-9762-4c1a-bf29-74b3cdfed8b6" />

Selection and hover over a tree node:
<img alt="Image" src="https://github.com/user-attachments/assets/75457f5f-8497-42d4-8daa-7be441fbe7fe" />

Nav cursor and selectable rounding in a menu:
<img alt="Image" src="https://github.com/user-attachments/assets/8c4d10da-d972-4e8d-846f-25fe9bd19be6" />

**Comment by LxeC** [comment]
This all looks very promising, I wonder if it's possible to have an optional amount of pixels or % of margin/padding around the items and borders, because the reason a lot of these cases with the rounding might look weird is because the rounded highlight/element is touching the border or the menu items are very closed to each other. Usually in design it looks much better if you give them some air to breathe,

<img width="750" height="185" alt="Image" src="https://github.com/user-attachments/assets/fc09c742-4c4d-4072-b771-aa39f12a26dc" />

<img width="944" height="471" alt="Image" src="https://github.com/user-attachments/assets/9d1581bd-8748-4ea1-811b-42d45df70d7b" />

This is part of the Gestalt principles in design, where spacing helps guide the eye better and group similar elements together more efficiently as well as psychological reasons as to why seeing very parallel near-overlapping lines feels annoying and "wrong" to the eye.

**Comment by janitor-raus** [comment]
I imagine this would be a matter of adjusting the inner item spacing for such cases when laying items out in grids. Menus probably need child/popup padding instead. I'm also in the process of exploring a few edge cases in relation to that.

Edit: also been messing around a bit with the MenuItem/BeginMenu, as I noticed the checkmark/arrow were right up against the rightmost edge of the widget, and the shortcut too close to the text by default:

<img alt="Image" src="https://github.com/user-attachments/assets/fbd63672-f648-49a5-9c7b-59cd47d28c39" />

And yes, something is weird with their outlines now. Seems to come from the "AA lines use textures" option. Wonder how/why that broke all of a sudden, or maybe I just never noticed..

Edit 2: I also just noticed that the menu bar does not properly take account of padding. My buttons are squished up against the top border of the menubar, and if the vertical frame padding is small enough, the bottom of the buttons is clipped off (easier to see with rounding as it won't show). I'll also need to look into and patch that.

**Comment by ocornut** [comment]
> This all looks very promising, I wonder if it's possible to have an optional amount of pixels or % of margin/padding around the items and borders, because the reason a lot of these cases with the rounding might look weird is because the rounded highlight/element is touching the border or the menu items are very closed to each other. Usually in design it looks much better if you give them some air to breathe,

`Selectable()` by design are intentionally covering the expected space between two items in order to leave no "clicking" gap.
You could argue this was not a good design and it should instead be reformulated into a way for users to specify more spacing options, which will require some clever design for backward compat (I have a few ideas) **but this is unrelated to this topic**. Please both of you don't spam the rounding topic with other ideas. Each of those ideas I can hear but they tend to need deeper design and discussion.

I'm honestly terrified by the amount of people trying to push dear imgui toward super neat look. I believe you are vastly underestimating the complexity/cost trade-off you are paying for, plus the effect it has on development velocity. Out of experience the smallest visual changes can have complexity effects on many things. I've seen hundreds of quick hacks and it's great that you can make them locally, but they have a tendency to have some other side-effect, so each of those change tend to warrant their own discussion to be accepted into main line. I see almost ten discussion topics casually mentioned in the last few messages. 

As for e.g. " List boxes are also alright: "  I don't think that's an alright look at all, but the other solution that keeps rounding is way too complex.
As for e.g. "as I noticed the checkmark/arrow were right up against the rightmost edge of the widget" this is not what I am seeing. Which suggest (over and over again) that you are not providing enough information, and this is why topics shouldn't be mixed up.

**Comment by LxeC** [comment]
Personally, I don't think it was spamming so much as it was just relevant suggestions. I am sorry if it got perceived the wrong way.

There's nothing wrong for making dear imgui have a professional "super neat look". I know this library started off a merely as a debugging tool with crude basic visuals but it can be so much more, it's such a missed opportunity. And it'd be nice to have this as much out of the box as possible.

As for performance, I don't think a few rounded corners and spacing options would be _that_ much of a performance hit or maintenance issue if the code is planned properly. If anything, couldn't it be as an optional module or fork of sorts? The dockable functionality is and I'd argue that'd be a bigger performance hit than some rounded corners and spacing, but again it is _optional_. Additionally, a _lot_ of imgui widgets are not really necessary for most apps, but they are still present, and a lot of people are thankful for that, I'd again argue these widgets are more noise than a few visual adjustments.

**Comment by ocornut** [comment]
> Personally, I don't think it was spamming so much as it was just relevant suggestions.

It is only spamming because it is off-topic here. If you want to discuss the padding of this or that, open another topic and provide relevant details, I'm happy to discuss this. But they have nothing to do with selectable/menu item rounding.. Each of those sub-topics which might look minor to you probably aren't. 

> There's nothing wrong for making dear imgui have a professional "super neat look"

I didn't say it was wrong. I said you are underestimating all the side effects this is causing to both dear imgui development and your own development. Each of those push for "more style variables!!" is adding complexity/cost and is hindering my freedom to make actual larger more meaningful changes on the styling system. We can't have everything. My job here is to make this technology last for a long time and be reliable and not fall apart under its own weight. Things are improving and will be improving but I am super careful.

**Comment by ocornut** [comment]
Basically if I expose a "SelectableRounding" option then I will get 10+ more requests trickling down from that change. 

(1) e.g. how do you make this happen? 
<img width="324" height="258" alt="Image" src="https://github.com/user-attachments/assets/2e47ecc4-63c6-429f-a1c6-7aa28d967460" />
Don't you think people will adjust a style value and then complain that it doesn't look right with rounding?

(2) It's also basically breaking my solution for a generic BeginSelectable/EndSelectable() scoped API which would typically rely on preallocating and reshading 4 vtx. 

(3) How do we deal with inconsistency with e.g. whole row spanning not exhibiting the same rounding?

People ask about things because they look nice, but ignore my messages stating what the unexpected cost might be.

1.93.x will luckily make rounded shapes cheaper in most scenarios (immense amount of work went into this, thanks to Mikko).

**Comment by ocornut** [comment]
> It is only spamming because it is off-topic here. If you want to discuss the padding of this or that, open another topic and provide relevant details, I'm happy to discuss this. But they have nothing to do with selectable/menu item rounding.. Each of those sub-topics which might look minor to you probably aren't.

I should also stress that posting USEFUL STUFF in the WRONG TOPIC is rather stressful for me in terms of issues/topics management. It essentially makes it difficult for me to locate design discussions and to close the topic if there is a viable idea in it that hasn't been acted on.

**Comment by ocornut** [comment]
I have pushed ac6f968 which adds `style.MenuItemRounding`, `style.SelectableRounding`, `ImGuiStyleVar_MenuItemRounding`, `ImGuiStyleVar_SelectableRounding`. I have intentionally not exposed SelectableRounding in the demo and added comments to discourage people from casually changing it. The aim of this change is to primarily support this for Menu Items (which are the primary reasons for those issues being opened).

Other padding/rendering/tweaks, if any, may be discussed in new issues. I'm happy to discuss them.

> Tree-specific RenderFrame() call aside, it turned out I had to remove the NoRounding flag for the Nav cursor render for consistency of selection outlines (square vs the rounded frame). Actual patch for reference here: https://github.com/janitor-raus/CubeChip/blob/master/patches/imgui/selectable_rounding.patch

The link doesn't work anymore so I can't tell what you did, but note that 0e355af changed `RenderNavCursor()` to take expicit rounding value which seems more saner. Might even remove the "default" value for it.

**Comment by ocornut** [comment]
I've seen this thread quoted by people who seems to misunderstand the problems I am pointing at (probably not the people who posted here), so to clarify for the outside reader:

> (1) e.g. how do you make this happen? 
> <img alt="Image" width="324" height="258" src="https://private-user-images.githubusercontent.com/8225057/612574281-2e47ecc4-63c6-429f-a1c6-7aa28d967460.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODI0MTY0MzgsIm5iZiI6MTc4MjQxNjEzOCwicGF0aCI6Ii84MjI1MDU3LzYxMjU3NDI4MS0yZTQ3ZWNjNC02M2M2LTQyOWYtYTFjNi03YWEyOGQ5Njc0NjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDYyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA2MjVUMTkzNTM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGE5YzE3Y2RjZTQwMDFlOTQxZDFhZDdhMjk3ODA4MjE5ODEwOTgwZjEzMDQxYTU4N2VlMjFhODQ4MTg3MmE3OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmcmVzcG9uc2UtY29udGVudC10eXBlPWltYWdlJTJGcG5nIn0.fPCmquiP2G33KRyJXQqYGSp0XB4DOBfRXGwtbWevEzs"> 

If you apply rounding to individual elements it won't look like this. 
This exact look, in a system that favor stateless submissions, requires a two-pass system for recording contiguous selection spans and drawing them when a span ends. The easiest way would be to use a ImDrawList splitter with a background draw channel (similar to what tables are done) in the `BeginMultiSelect()` scope. It's technically possible to solve fairly efficiently but requires some work, and there are inevitable complications if using a clipper. I think we can probably attempt this work at the same time as I refactor the way tables draw cell borders and backgrounds, e.g. storing them in a queue. 

In the meanwhile, if you were use rounded Selectable with span of selected items it won't look right.

Rounded menus are fine because they are not simultaneously highlighted, so you can use them today without issue. 

> (2) It's also basically breaking my solution for a generic BeginSelectable/EndSelectable() scoped API which would typically rely on preallocating and reshading 4 vtx.

Perhaps we should generalize the concept of a background draw channel in all windows.

TL;DR; out of experience I just intuitively know that making this option a first-class citizen with inevitably both raise new issues and raise the bars further for more styling variables. I'm always reluctant to add more styling variables in the current styling system because it is not well designed enough, and every variables will be an additional burden when it comes to designing a more general/better system.

> I'd again argue these widgets are more noise than a few visual adjustments.

The stuff proposed in your mockups may appear as just "a few visual adjustments", but the thing you are removing from the pictures have other uses and purposes in other situations. There are situation where UI can't afford having clicking gap, there are situations where UI absolutely needs clicking gaps (e.g. to start a box-select in a context where drag and drop is also used on items). There are situation where we want an intentional mismatch between visuals and hit-boxes. The whole design of `Selectable()` taking over 0.5f*ItemSpacing needs to changed (it has caused many problems and was not a great idea). Those are perfectly fine topics to point out and solve, and I've worked hours on each of them already. I was just outlining, they are NOT simple/minor changes, they have huge implications on users, and therefore they shouldn't casually be mentioned in a random thread while passing by.
